File: oldScratch/cpp/take2/MaxGroup.cpp
#include "MaxGroup.h"
#include "FastCalcTile.h"
//max groups creation is not performance critical, so check constraints
MaxGroup::MaxGroup(const std::vector<Tile*>& tilesFound){
    for(unsigned int i=0;i<tilesFound.size();i++){
        allGroup.push_back(tilesFound[i]->toFastCalcTile());
    }
    selected=0;

    //if 3, all or nothing,
    //if 4, all, nothing, and the 4 3-groups leaving out 1
    //if 6, same as 4 but also the possibility of 2 3-groups at the same time
    possibilityCount=7;
    if(allGroup.size()==3) {
        possibilityCount=2;
    }
    else if(allGroup.size()==4){
        possibilityCount=6;
    }
    //validate
    std::stringstream err;
    err<<"invalid group [";
    if(tilesFound.size()>0){
        err<<tilesFound[0]->debugString();
        for(unsigned int i=0;i<tilesFound.size();i++){
            err<<","<<tilesFound[i];
        }
    }
    err<<"]: ";
    if(allGroup.size()!=3&&allGroup.size()!=4&&allGroup.size()!=6){
        err<<"size must be either 3, 4, or 6";
        throw err.str();
    }
    int max=tilesFound.size();
    if(max>4) max=4;//assert the first 3 or 4 tiles are sequential,
    //forcing the duplicate to the last 2
    int num=tilesFound[0]->getNumber();
    TileColor lastColor=tilesFound[0]->getTileColor();
    std::string error="";
    for(int i=1;i<max;i++){
        if(tilesFound[i]->getNumber()!=num){
            error="numbers must match";
            break;
        }
        if(tilesFound[i]->getTileColor()<=lastColor){
            error="the non duplicate colors must be sequential";
            break;
        }
        lastColor=tilesFound[i]->getTileColor();
    }
    if(error!=""){
        err<<error;
        throw err.str();
    }
}
std::vector<std::vector<unsigned int>> MaxGroup::getGroupForPossibilityKey(unsigned int possibilityKey){
    //TODO: this hasn't been updated to handle the 6 group
    std::vector<unsigned int> group;
    //last is all used
    if(possibilityKey==possibilityCount-1){
        group=allGroup;
    }
    else if(possibilityKey!=0){
        for(unsigned int i=0;i<allGroup.size();i++){
            //omit the selected which went in the unused
            if(i!=possibilityKey-1){
                group.push_back(allGroup[i]);
            }
        }
    }
    //else key==0, 0 is all unused, so return empty group
    std::vector<std::vector<unsigned int>> result;
    result.push_back(group);
    return result;
}
void MaxGroup::addCurrentUnused(int addTo[],int& addLocation){
    if(selected==0){
        for(unsigned int i=0;i<allGroup.size();i++){
            addTo[addLocation]=allGroup[i];
            addLocation++;
        }
    }
    else{
        //if(possibilityCount==2) done, using all tiles
        //1-4= return that tile as unused
        //else possibility #6: selected==5 and so using all the tiles, no add
        if(selected<5){
            addTo[addLocation]=allGroup[selected-1];
            addLocation++;
        }
        //selected==5 same as 4 tile case (using the first 4), but add the two dups to unused
        //if we have the 2 dups add them unless all used
        if(possibilityCount==7&&selected!=6){
            addTo[addLocation]=allGroup[4];
            addLocation++;
            addTo[addLocation]=allGroup[5];
            addLocation++;
        }
    }
}
unsigned int MaxGroup::getCurrentPossibilityKey(){
    return selected;
}
bool MaxGroup::isAtLast(){
    return selected==possibilityCount-1;
}
void MaxGroup::resetIteration(){
    selected=0;
}
void MaxGroup::moveNext(){
    selected++;
}
//allow sorting by group number
bool MaxGroup::operator<(const MaxGroup& other){
    //groups have all the same number;
    return allGroup[0]<other.allGroup[0];
}
bool MaxGroup::pointerLess(MaxGroup* r,MaxGroup* l)
{
    return *r<*l;
}
//unsigned int MaxGroup::allGroupSize(){
    //return allGroup.size();
//}
unsigned int MaxGroup::getPossibilityCount(){
    return possibilityCount;
}
std::string MaxGroup::debugString(){
    std::stringstream s;
    s<<"("<<selected<<"/"<<possibilityCount-1<<")[";
    if(allGroup.size()>0)
    {
        s<<FastCalcTile::Printable(allGroup[0]).toString();
        for(unsigned int i=1;i<allGroup.size();i++){
            s<<","<<FastCalcTile::Printable(allGroup[i]).toString();
        }
    }
    s<<"]"<<std::endl;
    return s.str();
}

File: RummikubSolver/SolverLogic/Models/MaxGroup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SolverLogic.Models
{
    internal struct SizeAndExcludeList
    {
        public int size;
        public int[] excludes;
    }
    public class MaxGroup
    {
        public const int Possibilities_3G = 2;
        public const int Possibilities_4G = 6;
        public const int Possibilities_6G = 7;
        private static readonly SizeAndExcludeList[][] SizeAndExcludeMapByPossibilitySelected;
        static MaxGroup()
        {
            SizeAndExcludeMapByPossibilitySelected = new SizeAndExcludeList[8][];
            SizeAndExcludeMapByPossibilitySelected[Possibilities_3G] = new SizeAndExcludeList[]
            {
                new SizeAndExcludeList{ size = 0, excludes = new int[] {0,1,2}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {}},
            };
            SizeAndExcludeMapByPossibilitySelected[Possibilities_4G] = new SizeAndExcludeList[]
            {
                new SizeAndExcludeList{ size = 0, excludes = new int[] {0,1,2,3}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {0}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {1}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {2}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {3}},
                new SizeAndExcludeList{ size = 4, excludes = new int[] {}}
            };
            SizeAndExcludeMapByPossibilitySelected[Possibilities_6G] = new SizeAndExcludeList[]
            {
                new SizeAndExcludeList{ size = 0, excludes = new int[] {0,1,2,3,4,5}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {0,4,5}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {1,4,5}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {2,4,5}},
                new SizeAndExcludeList{ size = 3, excludes = new int[] {3,4,5}},
                new SizeAndExcludeList{ size = 4, excludes = new int[] {4,5}},
                new SizeAndExcludeList{ size = 6, excludes = new int[] {}}
            };
        }
        private readonly List<Tile> allGroup;
        public int PossibilityCount { get; private set; }
        private BitVectorPerColor[] orsPerPossibility;
        public MaxGroup(List<Tile> tilesFound)
        {
            allGroup = tilesFound;
            //if 3, all or nothing,
            //if 4, all, nothing, and the 4 3-groups leaving out 1
            //if 6, same as 4 but also the possibility of 2 3-groups at the same time
            PossibilityCount = Possibilities_6G;
            if (allGroup.Count== 3)
            {
                PossibilityCount = Possibilities_3G;
            }
            else if (allGroup.Count == 4)
            {
                PossibilityCount = Possibilities_4G;
            }
            string errorCtx = $"invalid group [{string.Join(",", tilesFound.Select(x => x.DebugDisplay))}]: ";
            if(!new[] { 3, 4, 6 }.Contains(allGroup.Count))
            {
                throw new ArgumentException(errorCtx+"size must be either 3, 4, or 6");
            }
            int max = tilesFound.Count;
            if (max > 4) max= 4;
            int num = tilesFound[0].Number;
            var lastColor=tilesFound[0].Color;
            for(int i = 1; i < max; i++)
            {
                if(tilesFound[i].Number != num)
                {
                    throw new ArgumentException(errorCtx+"numbers in groups must match");
                }
                if(tilesFound[i].Color <= lastColor)
                {
                    throw new ArgumentException(errorCtx+"the non duplicate colors must be sequential");
                }
                lastColor = tilesFound[i].Color;
            }
            orsPerPossibility = new BitVectorPerColor[PossibilityCount];
            for(int key = 0; key < PossibilityCount; key++)
            {
                var res=SizeAndExcludeMapByPossibilitySelected[PossibilityCount][key];
                for(int i = 0; i < allGroup.Count; i++)
                {
                    if(res.excludes.Contains(i))
                    {
                        orsPerPossibility[key].SetColorBit(allGroup[i].Color, allGroup[i].CanonicalIndex, true);
                    }
                }
            }
        }
        public Tile[] GetGroupForPossibilityKey(int key)
        {
            var res=SizeAndExcludeMapByPossibilitySelected[PossibilityCount][key];
            if (res.size == 0) return new Tile[] { };
            var group = new Tile[res.size];
            int index = 0;
            for(int i = 0; i < allGroup.Count; i++)
            {
                if(!res.excludes.Contains(i))
                {
                    group[index] = allGroup[i];
                    index++;
                }
            }
            return group;
            
        }
        public void MarkUnusedForSelected(ref UnusedTilesState unusedTiles, int key)
        {
            unusedTiles.UnusedInGroupsFlags.OrEquals(ref orsPerPossibility[key]);
        }
        public static IComparer<MaxGroup> Comparer { get; } = Comparer<MaxGroup>.Create((x, y) => x.allGroup[0].Number - y.allGroup[0].Number);
        public override string ToString()
        {
            var reps=new List<string>();
            foreach(var tile in allGroup)
            {
                reps.Add(tile.ToString());
            }
            return $"{PossibilityCount}[{string.Join(",", reps)}]";
        }
    }
}
